test/tools/isoltest.cpp

```diff
@@ -0,0 +1,326 @@
+		FormattedPrinter(_colored), m_name(_name), m_path(_path)
+	{}
+
+	bool process(int& _successCount);
+
+	static bool processPath(
+		fs::path const& _basepath,
+		fs::path const& _path,
+		int& _successCount,
+		int& _runCount
+	);
+
+	static string editor;
+	static bool noColor;
+private:
+	bool handleResponse(int& _successCount, bool const _parserError);
+
+	void printContract() const;
+
+	string const m_name;
+	fs::path const m_path;
+	unique_ptr<SyntaxTest> m_test;
+};
+
+bool SyntaxTestTool::noColor = false;
+string SyntaxTestTool::editor;
+
+void SyntaxTestTool::printContract() const
+{
+	stringstream stream(m_test->source());
+	string line;
+	Scope formattedStream = format(cout, {CYAN});
+	while (getline(stream, line))
+		formattedStream << "    " << line << endl;
+	stream << endl;
+}
+
+bool SyntaxTestTool::process(int& _successCount)
+{
+	bool success;
+	bool parserError = false;
+	std::stringstream outputMessages;
+
+	format(cout, {BOLD}) << m_name << ": ";
+	cout.flush();
+
+	try
+	{
+		m_test = unique_ptr<SyntaxTest>(new SyntaxTest(m_path.string(), !noColor));
+	}
+	catch (std::exception const& _e)
+	{
+		format(cout, {RED}) << "cannot read test: " << _e.what() << endl;
+		return true;
+	}
+
+	try
+	{
+		success = m_test->run(outputMessages, "  ", 2);
+	}
+	catch (...)
+	{
+		success = false;
+		parserError = true;
+	}
+
+	if (success)
+	{
+		format(cout, {GREEN}) << "OK";
+		cout << endl;
+		++_successCount;
+		return true;
+	}
+	else
+	{
+		format(cout, {RED}) << "FAIL";
+		cout << endl;
+
+		cout << "  Contract:" << endl;
+		printContract();
+
+		if (parserError)
+		{
+			cout << "  ";
+			format(cout, {INVERSE, RED}) << "Parsing failed:" << endl;
+			m_test->printErrorList(cout, m_test->compilerErrors(), "    ", true, true);
+			cout << endl;
+		}
+		else
+			cout << outputMessages.str() << endl;
+
+		return handleResponse(_successCount, parserError);
+	}
+}
+
+bool SyntaxTestTool::handleResponse(int& _successCount, bool const _parserError)
+{
+	if (_parserError)
+		cout << "(e)dit/(s)kip/(q)uit? ";
+	else
+		cout << "(e)dit/(r)eplace/(s)kip/(q)uit? ";
+	cout.flush();
+
+	while (true)
+	{
+		switch(cin.get())
+		{
+		case 's':
+			cout << endl;
+			return true;
+		case 'r':
+			cout << endl;
+			if (!_parserError)
+			{
+				ofstream file(m_path.string(), ios::trunc);
+				file << m_test->source();
+				file << "// ----" << endl;
+				if (!m_test->errorList().empty())
+				{
+					m_test->disableFormatting();
+					m_test->printErrorList(file, m_test->errorList(), "// ", false, false);
+					if (!noColor)
+						m_test->enableFormatting();
+				}
+			}
+			return process(_successCount);
+		case 'e':
+			cout << endl << endl;
+			if (system((editor + " " + m_path.string()).c_str()))
+				cerr << "Error running editor command." << endl << endl;
+			return process(_successCount);
+		case 'q':
+			cout << endl;
+			return false;
+		default:
+			break;
+		}
+	}
+}
+
+
+bool SyntaxTestTool::processPath(
+	fs::path const& _basepath,
```

Comment:
Is it important to distinguish between `_basepath` and `_path` for this function?

Reply:
I would say that the output of isoltest is more readable, if it prints paths relative to the base path and not absolute paths -